### DISCUSSION TITLE
Conditional rules?

### DISCUSSION THREAD

**Discussion by juh9870** [TodePond/CellPond#289] [Help]
Is there a way to make a rule only apply if some other rule failed? Instead of being random.

**Comment by TodePond** [comment]
Sorry no! This is what the unused 'down-pointing-triangle' will eventually be used for. But that would involve someone else making it, or me making another video about the engine, so probably not any time soon

**Reply by raisedeyebrowemoji** [reply]
So something like this?
![2023-08-22_08-43-04](https://github.com/TodePond/CellPond/assets/113126734/560047d9-1cd5-4d0b-a057-cd39a6776dd8)
The white cell should fall until it no longer can and then should dissapear?

**Reply by TodePond** [reply]
Close! But not exactly.

[MarkovJunior](#) works a bit like that. It's possible in MarkovJunior because the way that it applies rules is different to CellPond. In MarkovJunior, it does this:

- pick a rule
- find all the places it could apply
- pick a random place
- apply the rule in that place

CellPond, on the other hand, does this:

- pick a place
- find all the rules that could apply here
- pick a random rule
- apply the rule in that place

So do you see how it's the reverse order?
- MarkovJunior: rule before place
- CellPond: place before rule

## So what?

With MarkovJunior, you can have a situation where you "run out of places" to apply a rule. If that's the case, you can defer to a later rule. But that situation doesn't happen in CellPond.

However, you **can** have a situation where you've tried to match a rule to the current place, and it doesn't fit. The rule might apply elsewhere, but with CellPond, we only care about the nearby surroundings (not the whole grid, like MarkovJunior).

So we **can** make a similar thing in CellPond, but it needs a small tweak.

## What's the tweak?

The silhouette of each row needs to be the same.

In MarkovJunior, when a rule fails, it tries to match the next rule in a brand new place, so the 'shape' of the rule can be different. eg:

```
❤️➡️💙
💙➡️❤️

❤️➡️💙
```

^ Note how the first rule is 1x2, but the second rule is 1x1.

With CellPond, we'd try to apply the second rule in the same place. This means we need some spatial consistency. In the example above, it's unclear: Should the second rule fire on the top cell? or the bottom cell?

Instead, we need to match the silhouettes of each rule:

```
❤️➡️💙
💙➡️❤️

❤️➡️💙
⭐➡️⭐
```

^ Note how both rules are 1x2.
PS: The star emoji is usually used to denote that the cell can be 'anything'.

## Is that it?

Yes that's it. The fun thing is that you could play around with the logic of it.

You can already turn rules 'off' by clicking the triangle within a rule. It toggles it between white and black.

You could use this to invert the logic of the rules. ie: You could choose between...

- If the rule DOESN'T fire, apply the second rule.
- If the rule DOES fire, apply the second rule.

Could be fun! I'll make it eventually
